### DISCUSSION TITLE
final PnL in pips incorrect?

### DISCUSSION THREAD

**Discussion by jjphung** [kernc/backtesting.py#269] [Q&A] [answered]
In an attempt to see PnL in terms of pips when dealing with forex, I stumbled upon this thread: https://github.com/kernc/backtesting.py/issues/8. In this thread @kernc suggested: `pnl_in_pips = self.position.pl / self.position.size / self.data.pip` in order to calculate PnL in terms of pips and I inserted into a strategy along the lines of:

```py
class BaseStrategy(Strategy)
    total_pips_gained = 0
    # ......    
    def next(self):
        super().next()
        exit_portion = self.__exit_signal[-1]
        print("Total pips gained: " ,self.total_pips_gained)
        if exit_portion > 0:
            for trade in self.trades:
                if trade.is_long:
                    pnl_in_pips = self.position.pl / self.position.size / self.data.pip
                    self.total_pips_gained += pnl_in_pips
                    trade.close(exit_portion)
        elif exit_portion < 0:
            for trade in self.trades:
                if trade.is_short:
                    pnl_in_pips = self.position.pl / self.position.size / self.data.pip
                    self.total_pips_gained += pnl_in_pips
                    trade.close(-exit_portion)
```
With the idea that we'd want to calculate PnL of pips when we close our position.

However, `stats = bt.run()` outputs `Return [%]: 1.6968`, while the last print statement for `total_pips_gained` prints out `-1033.8971649107916`. Am calculating pnl_in_pips correctly? How is it that return % overall is positive, but total_pips_gained is very negative.

UPDATE:

For shorts: it should be: `self.total_pips_gained += (pnl_in_pips * -1)`? Since it is the opposite direction! Is this logic correct?

Follow up question: I was not able to find `self.data.pip` anywhere in the codebase on how it is determined. My forex data is in OHCLV format. 

Thank you!

**Comment by kernc** [comment, marked as answer]
Not sure it's related, but you likely _should_ reset `total_pips_gained` in `init()`:
```py
    def init(self):
        self.total_pips_gained = 0
```
Let me know if that does anything.



> I was not able to find `self.data.pip` anywhere in the codebase on how it is determined.

It's here:
https://github.com/kernc/backtesting.py/blob/5e1acccb7b0c9c1f369531035a9c887736f92cc5/backtesting/_util.py#L141-L146

**Reply by jjphung** [reply]
Thank you!

> Let me know if that does anything.

It does not.

**Reply by kernc** [reply]
What about if you [stop dividing by `self.position.size`](https://github.com/kernc/backtesting.py/issues/8#issuecomment-508555107)?

**Reply by jjphung** [reply]
Interesting. If I don't divide by `self.position.size`, pips become an absurd number.. Each trade is in the multi-millions on the EUR_USD pair.

Ex:
```
SELL
Total pips gained:  -8888523.600199785
```

**Comment by kernc** [comment]
Let's extend the simple SmaCross strategy from the tests:
https://github.com/kernc/backtesting.py/blob/73e1534428fc2db9e681030862133b5067508520/backtesting/test/_test.py#L54-L69

We override it such that `self.position.close()` calls that the strategy uses also extract and save `pnl_in_pips`:
```py
from backtesting import Backtest, Strategy
from backtesting.test import GOOG
from backtesting.test._test import SmaCross

class S(SmaCross):
    def init(self):
        super().init()        
        self.total_pips = 0
        
        # Patch self.position.close()
        old_position_close = self.position.close
        def new_position_close(*args, **kwargs):
            if self.position:
                pnl_in_pips = self.position.pl / self.data.pip
                print(pnl_in_pips, self.trades[-1])
                self.total_pips += pnl_in_pips
            old_position_close(*args, **kwargs)
        self.position.close = new_position_close
            
bt = Backtest(GOOG.iloc[:100], S, cash=10_000, trade_on_close=True)
stats = bt.run()
stats._strategy.total_pips
stats._trades
stats[['Equity Final [$]', 'Return [%]']]
```
With this, I get:
![Screenshot_2021-03-08_03-52-04](https://user-images.githubusercontent.com/684364/110269210-87e1bb00-7fc3-11eb-9308-8d22a1c2f364.png)

The numbers all add up. (There is one additional trade [forcefully closed at the end of the backtest run](https://github.com/kernc/backtesting.py/blob/73e1534428fc2db9e681030862133b5067508520/backtesting/backtesting.py#L1167-L1169), and that closing doesn't call our patched method, hence the missing output.)

Notice the use of [`trade_on_close=True`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest). Without it, I get a slightly different output:

![Screenshot_2021-03-08_04-02-54-2](https://user-images.githubusercontent.com/684364/110269321-c0819480-7fc3-11eb-8756-8b1ae81c2464.png)

a result of the fact trades aren't closed instantly, but rather with the _next bar's open_.
`bt.plot()` confirms.

**Reply by jjphung** [reply]
Wow thank you for your dedication to answer my question and clarifying. Much appreciated!

**Comment by kernc** [comment]
> For shorts: it should be: `self.total_pips_gained += (pnl_in_pips * -1)`? Since it is the opposite direction! Is this logic correct?

Best just look in the source or in [docs](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Trade.pl):
https://github.com/kernc/backtesting.py/blob/73e1534428fc2db9e681030862133b5067508520/backtesting/backtesting.py#L600-L604

`size` (volume) is the only attribute whose sign corresponds with direction, afaik. And yes, you are using it. :grin:
